crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -810,54 +810,84 @@ impl<'src> Parser<'src> {
         }
     }
 
-    /// See: <https://docs.python.org/3/reference/simple_stmts.html#grammar-token-python-grammar-assignment_stmt>
+    /// Parse an assignment statement.
+    ///
+    /// # Panics
+    ///
+    /// If the parser isn't positioned at an `=` token.
+    ///
+    /// See: <https://docs.python.org/3/reference/simple_stmts.html#assignment-statements>
     fn parse_assign_statement(&mut self, target: ParsedExpr, start: TextSize) -> ast::StmtAssign {
         let mut targets = vec![target.expr];
-        let mut value = self.parse_expression_list();
 
-        if self.at(TokenKind::Equal) {
-            self.parse_list(RecoveryContextKind::AssignmentTargets, |parser| {
-                parser.bump(TokenKind::Equal);
+        // test_err assign_stmt_missing_rhs
+        // x =
+        // 1 + 1
+        // x = y =
+        // 2 + 2
+        // x = = y
+        // 3 + 3
+        self.parse_list(RecoveryContextKind::AssignmentTargets, |parser| {
+            parser.bump(TokenKind::Equal);
+
+            // test_err_assign_stmt_keyword_target
+            // a = pass = c
+            // a + b
+            // a = b = pass = c
+            // a + b
+            targets.push(parser.parse_star_expression_list().expr);
+        });
 
-                let mut parsed_expr = parser.parse_expression_list();
+        // test_err assign_stmt_invalid_starred_expr
+        // x = *a and b
+        // x = *yield x
+        // x = *lambda x: x
 
-                std::mem::swap(&mut value, &mut parsed_expr);
+        // The last target is the value of the assignment statement.
+        // SAFETY: There is at least one element in the vector which is the parameter `target`.
+        let value = targets.pop().unwrap();
```

Comment:
Consuming the `=` token in the parsing function helps us remove the memory swap logic. What we do now is collect all the expression assuming that they are assignment targets and at the end pop one from the end which will become the value. Then, we'll validate the targets.